GPU/Common/GPUStateUtils.cpp

```diff
@@ -386,3 +408,627 @@ LogicOpReplaceType ReplaceLogicOpType() {
+		return;
+
+	case REPLACE_BLEND_COPY_FBO:
+		blendState.applyShaderBlending = true;
+		blendState.resetShaderBlending = false;
+		blendState.replaceAlphaWithStencil = replaceAlphaWithStencil;
+		break;
+
+	case REPLACE_BLEND_PRE_SRC:
+	case REPLACE_BLEND_PRE_SRC_2X_ALPHA:
+		usePreSrc = true;
+		break;
+
+	case REPLACE_BLEND_STANDARD:
+	case REPLACE_BLEND_2X_ALPHA:
+	case REPLACE_BLEND_2X_SRC:
+		break;
+	}
+
+	blendState.enabled = true;
+	blendState.resetShaderBlending = true;
+
+	const GEBlendMode blendFuncEq = gstate.getBlendEq();
+	GEBlendSrcFactor blendFuncA = gstate.getBlendFuncA();
+	GEBlendDstFactor blendFuncB = gstate.getBlendFuncB();
+	const u32 fixA = gstate.getFixA();
+	const u32 fixB = gstate.getFixB();
+
+	if (blendFuncA > GE_SRCBLEND_FIXA)
+		blendFuncA = GE_SRCBLEND_FIXA;
+	if (blendFuncB > GE_DSTBLEND_FIXB)
+		blendFuncB = GE_DSTBLEND_FIXB;
+
+	float constantAlpha = 1.0f;
+	BlendFactor constantAlphaGL = BlendFactor::ONE;
+	if (gstate.isStencilTestEnabled() && replaceAlphaWithStencil == REPLACE_ALPHA_NO) {
+		switch (ReplaceAlphaWithStencilType()) {
+		case STENCIL_VALUE_UNIFORM:
+			constantAlpha = (float)gstate.getStencilTestRef() * (1.0f / 255.0f);
+			break;
+
+		case STENCIL_VALUE_INCR_4:
+		case STENCIL_VALUE_DECR_4:
+			constantAlpha = 1.0f / 15.0f;
+			break;
+
+		case STENCIL_VALUE_INCR_8:
+		case STENCIL_VALUE_DECR_8:
+			constantAlpha = 1.0f / 255.0f;
+			break;
+
+		default:
+			break;
+		}
+
+		// Otherwise it will stay GL_ONE.
+		if (constantAlpha <= 0.0f) {
+			constantAlphaGL = BlendFactor::ZERO;
+		} else if (constantAlpha < 1.0f) {
+			constantAlphaGL = BlendFactor::CONSTANT_ALPHA;
+		}
+	}
+
+	// Shortcut by using GL_ONE where possible, no need to set blendcolor
+	bool approxFuncA = false;
+	BlendFactor glBlendFuncA = blendFuncA == GE_SRCBLEND_FIXA ? blendColor2Func(fixA, approxFuncA) : genericALookup[blendFuncA];
+	bool approxFuncB = false;
+	BlendFactor glBlendFuncB = blendFuncB == GE_DSTBLEND_FIXB ? blendColor2Func(fixB, approxFuncB) : genericBLookup[blendFuncB];
+
+	if (gstate.FrameBufFormat() == GE_FORMAT_565) {
+		if (blendFuncA == GE_SRCBLEND_DSTALPHA || blendFuncA == GE_SRCBLEND_DOUBLEDSTALPHA) {
+			glBlendFuncA = BlendFactor::ZERO;
+		}
+		if (blendFuncA == GE_SRCBLEND_INVDSTALPHA || blendFuncA == GE_SRCBLEND_DOUBLEINVDSTALPHA) {
+			glBlendFuncA = BlendFactor::ONE;
+		}
+		if (blendFuncB == GE_DSTBLEND_DSTALPHA || blendFuncB == GE_DSTBLEND_DOUBLEDSTALPHA) {
+			glBlendFuncB = BlendFactor::ZERO;
+		}
+		if (blendFuncB == GE_DSTBLEND_INVDSTALPHA || blendFuncB == GE_DSTBLEND_DOUBLEINVDSTALPHA) {
+			glBlendFuncB = BlendFactor::ONE;
+		}
+	}
+
+	if (usePreSrc) {
+		glBlendFuncA = BlendFactor::ONE;
+		// Need to pull in the fixed color. TODO: If it hasn't changed, no need to dirty.
+		if (blendFuncA == GE_SRCBLEND_FIXA) {
+			blendState.dirtyShaderBlend = true;
+		}
+	}
+
+	if (replaceAlphaWithStencil == REPLACE_ALPHA_DUALSOURCE && gstate_c.Supports(GPU_SUPPORTS_DUALSOURCE_BLEND)) {
```

Comment:
So I guess GPU_SUPPORTS_DUALSOURCE_BLEND is false on D3D9, right?  This won't have OpenGL data in it, I think...

-[Unknown]